AOTAutograd: correctness fix when tracing custom autograd functions that alias inputs

bdhirsh opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/102970. See the comment [here](https://github.com/pytorch/pytorch/issues/102970#issuecomment-1577223773) for details.

We normally treat "outputs that alias inputs" specially in AOTAutograd, by replaying the views at runtime, instead of baking them into the graph. For views that are part of custom autograd functions though, we can't do that view-replay, since it will clobber the backwards function that the user specified in their custom autograd.Function.

Right now in this PR, I distinguish between "aliased inputs that are normal views" vs. "aliased inputs that are views that came from an autograd.Function call" by checking the outputs `.grad_fn` field, to see if it inherits from our custom CBackward function class. Then I added a new `OutputType` enum value, that we effectively treat the "normal" way (the same way that we treat ordinary, non-aliased outputs). The new enum val is mostly for debugging - so we can print it and know that our graph had custom autograd.Function aliased outputs in it.

cc @zou3519 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106315
* __->__ #102992

bdhirsh committed:
AOTAutograd: correctness fix when tracing custom autograd functions that alias inputs

[ghstack-poisoned]

ezyang commented on test/functorch/test_aotdispatch.py:
```diff
@@ -551,6 +551,23 @@ def create_inp(req_grad):
         self.verify_aot_autograd(f, create_inp(True), test_mutation=True)
         self.verify_aot_autograd(f, create_inp(False), test_mutation=True)
 
+    def test_input_output_aliase_custom_autograd_function(self):
```
nit: alias

ezyang reviewed (approved):
This patch seems fine. The more involved version I want to see if we can scope out tomorrow involves exhaustively categorizing all of the possible states a `grad_fn` can be and planning out what we should do in each case.

zou3519 reviewed (approved):
This PR looks good to me, but we should test (and also fix, if possible and appropriate) the situation with C++ torch::autograd::Function. This is because people commonly use C++ torch::autograd::Function to register autograd kernels for custom operators and the case where one returns the input as-is is relatively common

bdhirsh commented:
Update - I have a C++ autograd::Function test that's failing locally. Still investigating

bdhirsh committed:
Update on "AOTAutograd: correctness fix when tracing custom autograd functions that alias inputs"

Fixes https://github.com/pytorch/pytorch/issues/102970. See the comment [here](https://github.com/pytorch/pytorch/issues/102970#issuecomment-1577223773) for details.

We normally treat "outputs that alias inputs" specially in AOTAutograd, by replaying the views at runtime, instead of baking them into the graph. For views that are part of custom autograd functions though, we can't do that view-replay, since it will clobber the backwards function that the user specified in their custom autograd.Function.

Right now in this PR, I distinguish between "aliased inputs that are normal views" vs. "aliased inputs that are views that came from an autograd.Function call" by checking the outputs `.grad_fn` field, to see if it inherits from our custom CBackward function class. Then I added a new `OutputType` enum value, that we effectively treat the "normal" way (the same way that we treat ordinary, non-aliased outputs). The new enum val is mostly for debugging - so we can print it and know that our graph had custom autograd.Function aliased outputs in it.

cc zou3519 




[ghstack-poisoned]

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -824,6 +829,11 @@ def inner(*flat_args):
+            elif curr_storage in inp_storage_refs and o.grad_fn is not None \
+                    and type(o.grad_fn).__name__ in ["CppFunction", "BackwardCFunction"]:
+                output_type = OutputType.custom_function_view
+                base_idx = None
```
cc @zou3519 finally updated and added some C++ tests (I ended up co-opting the open registration test file, since I found it very convenient for authoring custom ops in C++ that I could run torch.compile on).

Good catch mentioning that we should check for C++ autograd functions that alias too - they needed explicit handling (because cpp and python grad_fn's have different python types).

zou3519 replied:
Nice!

bdhirsh committed:
Update on "AOTAutograd: correctness fix when tracing custom autograd functions that alias inputs"

Fixes https://github.com/pytorch/pytorch/issues/102970. See the comment [here](https://github.com/pytorch/pytorch/issues/102970#issuecomment-1577223773) for details.

We normally treat "outputs that alias inputs" specially in AOTAutograd, by replaying the views at runtime, instead of baking them into the graph. For views that are part of custom autograd functions though, we can't do that view-replay, since it will clobber the backwards function that the user specified in their custom autograd.Function.

Right now in this PR, I distinguish between "aliased inputs that are normal views" vs. "aliased inputs that are views that came from an autograd.Function call" by checking the outputs `.grad_fn` field, to see if it inherits from our custom CBackward function class. Then I added a new `OutputType` enum value, that we effectively treat the "normal" way (the same way that we treat ordinary, non-aliased outputs). The new enum val is mostly for debugging - so we can print it and know that our graph had custom autograd.Function aliased outputs in it.

cc zou3519 




[ghstack-poisoned]

zou3519 reviewed (approved):
thank you!

bdhirsh committed:
Update on "AOTAutograd: correctness fix when tracing custom autograd functions that alias inputs"

Fixes https://github.com/pytorch/pytorch/issues/102970. See the comment [here](https://github.com/pytorch/pytorch/issues/102970#issuecomment-1577223773) for details.

We normally treat "outputs that alias inputs" specially in AOTAutograd, by replaying the views at runtime, instead of baking them into the graph. For views that are part of custom autograd functions though, we can't do that view-replay, since it will clobber the backwards function that the user specified in their custom autograd.Function.

Right now in this PR, I distinguish between "aliased inputs that are normal views" vs. "aliased inputs that are views that came from an autograd.Function call" by checking the outputs `.grad_fn` field, to see if it inherits from our custom CBackward function class. Then I added a new `OutputType` enum value, that we effectively treat the "normal" way (the same way that we treat ordinary, non-aliased outputs). The new enum val is mostly for debugging - so we can print it and know that our graph had custom autograd.Function aliased outputs in it.

cc zou3519 




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "AOTAutograd: correctness fix when tracing custom autograd functions that alias inputs"

Fixes https://github.com/pytorch/pytorch/issues/102970. See the comment [here](https://github.com/pytorch/pytorch/issues/102970#issuecomment-1577223773) for details.

We normally treat "outputs that alias inputs" specially in AOTAutograd, by replaying the views at runtime, instead of baking them into the graph. For views that are part of custom autograd functions though, we can't do that view-replay, since it will clobber the backwards function that the user specified in their custom autograd.Function.

Right now in this PR, I distinguish between "aliased inputs that are normal views" vs. "aliased inputs that are views that came from an autograd.Function call" by checking the outputs `.grad_fn` field, to see if it inherits from our custom CBackward function class. Then I added a new `OutputType` enum value, that we effectively treat the "normal" way (the same way that we treat ordinary, non-aliased outputs). The new enum val is mostly for debugging - so we can print it and know that our graph had custom autograd.Function aliased outputs in it.

cc zou3519 




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
